src/queryHistory/queryHistoryProvider.ts

```diff
@@ -0,0 +1,206 @@
+                (a as QueryHistoryNode).timeStamp.getTime();
+        });
+        this._onDidChangeTreeData.fire();
+    }
+
+    getTreeItem(node: QueryHistoryNode): any {
+        return node;
+    }
+
+    async getChildren(element?: any): Promise<vscode.TreeItem[]> {
+        if (this._queryHistoryNodes.length == 0) {
+            this._queryHistoryNodes.push(new EmptyHistoryNode());
+        }
+        return this._queryHistoryNodes;
+    }
+
+    /**
+     *
+     */
+    public async showQueryHistoryCommandPalette(): Promise<void> {
+        const options = this._queryHistoryNodes.map(node => this._queryHistoryUI.convertToQuickPickItem(node));
+        let queryHistoryQuickPickItem = await this._queryHistoryUI.showQueryHistoryCommandPalette(options);
+        this.openQueryHistoryEntry(queryHistoryQuickPickItem.node, queryHistoryQuickPickItem.action ===
+            QueryHistoryAction.RunQueryHistoryAction);
+    }
+
+    /**
+     * Starts the history capture by changing the setting
+     * and changes context for menu actions
+     */
+    public async startQueryHistoryCapture(): Promise<void> {
+        await this._vscodeWrapper.setConfiguration(Constants.extensionConfigSectionName,
+            Constants.configEnableQueryHistoryCapture, true);
+        await this._vscodeWrapper.setContext(Constants.isQueryHistoryEnabled, true);
+    }
+
+    /**
+     * Pauses the history capture by changing the setting
+     * and changes context for menu actions
+     */
+    public async pauseQueryHistoryCapture(): Promise<void> {
+        await this._vscodeWrapper.setConfiguration(Constants.extensionConfigSectionName,
+            Constants.configEnableQueryHistoryCapture, false);
+        await this._vscodeWrapper.setContext(Constants.isQueryHistoryEnabled, false);
+    }
+
+    /**
+     * Opens a query history listing in a new query window
+     */
+    public async openQueryHistoryEntry(node: QueryHistoryNode, isExecute: boolean = false): Promise<void> {
+        const editor = await this._untitledSqlDocumentService.newQuery(node.queryString);
+        let uri = editor.document.uri.toString(true);
+        let title = path.basename(editor.document.fileName);
+        const queryUriPromise = new Deferred<boolean>();
+        let credentials = this._connectionManager.getConnectionInfo(node.ownerUri).credentials;
+        await this._connectionManager.connect(uri, credentials, queryUriPromise);
+        await queryUriPromise;
+        this._statusView.languageFlavorChanged(uri, Constants.mssqlProviderName);
+        this._statusView.sqlCmdModeChanged(uri, false);
+        if (isExecute) {
+            const queryPromise = new Deferred<boolean>();
+            await this._outputContentProvider.runQuery(this._statusView, uri, undefined, title, queryPromise);
+            await queryPromise;
+            await this._connectionManager.connectionStore.removeRecentlyUsed(<IConnectionProfile>credentials);
```

Comment:
This again...it really would be great to make it so when you run the query you tell it whether you want to add it to the recently used list instead of relying on removing it manually each time.